### PR TITLE
Integration

### LINKED ISSUE #63: Integrate software components
As a daft engineer, I want to be able to use all the small bits of code we've written to actually have a vocoder.

Success criteria:

- Main thread that creates and owns threads for each task
- Attach message boxes to each thread and own locks for them

### PR DESCRIPTION
**Pull request by MarkAHarley**
Closes #63 

Notes to fix:

- RMS num samples should be larger for low frequency bands and smaller for high frequency bands
- RMS needs to incorporate a "microphone gain" field which is just a number to scale the output by
- Filter needs to Boost high frequency components (vocal sounds live there)

### CODE REVIEW

**Comment by MHCooke** [comment]
Do you intend to fix those notes in this branch or open them as tickets?

**Comment by MarkAHarley** [comment]
> Do you intend to fix those notes in this branch or open them as tickets?

Already wrote new tickets

**Review by Jambls — APPROVED** [review]

**Comment by MHCooke** [comment]
> Already wrote new tickets

Great, could you link them in the ticket please

**Review by bear-in-the-air — COMMENTED** [review]
May also want documented before merge

File: src/main.cc [hunk 1]
```diff
@@ -1,5 +1,87 @@
...
+    std::array<Pipe<Audio>, num_filters> modulator_out_pipes;
+
+    std::array<Pipe<double>, num_filters> power_out_pipes;
+
+    std::array<Pipe<Audio>, num_filters> amp_out_pipes;
+
+    Pipe<Audio> mixer_out_pipe;
+    mixer::Mixer<num_filters> mix(amp_out_pipes, mixer_out_pipe, timeout);
+
+    AlsaOut alsa_out("hw:2,0,0", mixer_out_pipe, timeout);
```
[the remark this answers is not in the document]
Comment (MarkAHarley) [reply src/main.cc:43]: future problem

File: src/main.cc [hunk 2]
```diff
@@ -1,5 +1,87 @@
...
+        // Filters
+        carrier_bank.emplace_back(2, alsa_in.get_sample_rate(), centre, width, carrier_in_pipes.at(i), carrier_out_pipes.at(i), timeout);
+        modulator_bank.emplace_back(2, alsa_in.get_sample_rate(), centre, width, modulator_in_pipes.at(i),
+                                    modulator_out_pipes.at(i), timeout);
+        centre += width;
+        carrier_bank.at(i).run();
+        modulator_bank.at(i).run();
+
+        // Power Meter
+        power_meter_bank.emplace_back(300, modulator_out_pipes.at(i), power_out_pipes.at(i), timeout);
```
[the remark this answers is not in the document]
Comment (MarkAHarley) [reply src/main.cc]: fixed

File: src/main.cc [hunk 3]
```diff
@@ -1,5 +1,87 @@
...
+    const int num_filters = 10;
+    const std::chrono::milliseconds timeout(2000);
+    std::vector<BPFilter> carrier_bank;
+    carrier_bank.reserve(num_filters);
+    std::vector<BPFilter> modulator_bank;
+    modulator_bank.reserve(num_filters);
+    int width = 10000 / num_filters; // NOLINT(cppcoreguidelines-avoid-magic-numbers)
+    int centre = width / 2;
+
+    std::vector<RMS> power_meter_bank; // NOLINT(cppcoreguidelines-avoid-magic-numbers)
```
[the remark this answers is not in the document]
Comment (MarkAHarley) [reply src/main.cc:25]: there was before RMS was changed, removed.

File: src/main.cc [hunk 4]
```diff
@@ -1,5 +1,87 @@
...
+
+    alsa_callback::acb alsa_in("hw:2,0,0");
+
+    const int num_filters = 10;
+    const std::chrono::milliseconds timeout(2000);
+    std::vector<BPFilter> carrier_bank;
+    carrier_bank.reserve(num_filters);
+    std::vector<BPFilter> modulator_bank;
+    modulator_bank.reserve(num_filters);
+    int width = 10000 / num_filters; // NOLINT(cppcoreguidelines-avoid-magic-numbers)
```
[the remark this answers is not in the document]
Comment (MarkAHarley) [reply src/main.cc:22]: it only symbolised that max freq is 10kHz, we could break it out to a constant but feels stupid as that's the only place it is used.

[the remark this answers is not in the document]
Comment (MHCooke) [reply src/main.cc:22]: Imo this is something that will be broken out in #125

File: src/main.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (MHCooke) [reply src/main.cc:43]: also #125

**Comment by MarkAHarley** [comment]
New issues created 
#124 
#128 
#133  
#132  
#129

File: include/alsa_out.hh [hunk 5]
```diff
@@ -44,7 +44,7 @@ class AlsaOut {
     // ALSA handle, modified in place by ALSA.
     snd_pcm_t *handle{};
     // Number of frames which ALSA reads/ writes at a time.
-    snd_pcm_uframes_t frames = 64; // NOLINT(cppcoreguidelines-avoid-magic-numbers) this is modified in place later
+    snd_pcm_uframes_t frames = 128; // NOLINT(cppcoreguidelines-avoid-magic-numbers) this is modified in place later
```
[the remark this answers is not in the document]
Comment (bear-in-the-air) [reply include/alsa_out.hh:47]: a comment noting that's what's going on would be good

File: src/main.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (bear-in-the-air) [reply src/main.cc:22]: that's what a magic number is. If it is going to be sorted later then I'm happy for it to be here for now

[the remark this answers is not in the document]
Comment (MarkAHarley) [reply src/main.cc:22]: cool then resolve this

**Review by MHCooke — APPROVED** [review]
Approving on the basis that a lot of the magic numbers will disappear with #125

**Review by MHCooke — CHANGES_REQUESTED** [review]
Can you remove the commented out print statements in the callback?

File: include/alsa_out.hh (hunk 5 above)
[the remark this answers is not in the document]
Comment (MarkAHarley) [reply include/alsa_out.hh:47]: done

**Review by MHCooke — APPROVED** [review]
Great!

**Pull request merged**